src/internet_identity/src/anchor_management/registration/registration_flow_v2.rs

```diff
@@ -0,0 +1,172 @@
+    else {
+        return Err(CheckCaptchaError::NoRegistrationFlow);
+    };
+
+    let RegistrationFlowState::CheckCaptcha {
+        flow_created_timestamp_ns,
+        captcha_solution,
+        ..
+    } = current_state
+    else {
+        return Err(CheckCaptchaError::UnexpectedCall {
+            next_step: RegistrationFlowNextStep::from(current_state),
+        });
+    };
+
+    let caller = caller();
+    if check_captcha_solution(arg.solution, captcha_solution).is_err() {
+        let (captcha_png_base64, flow_state) = captcha_flow_state(flow_created_timestamp_ns).await;
+        state::with_flow_states_mut(|flow_states| {
+            flow_states.update_registration_flow(caller, flow_state)
+        })
+        // If we fail to update the flow, then the flow has expired in the time it took to create
+        // the captcha.
+        .map_err(|_| CheckCaptchaError::NoRegistrationFlow)?;
+
+        return Err(CheckCaptchaError::WrongSolution {
+            new_captcha_png_base64: captcha_png_base64.0,
+        });
+    }
+
+    let flow_state = RegistrationFlowState::FinishRegistration {
+        flow_created_timestamp_ns,
+    };
+
+    state::with_flow_states_mut(|flow_states| {
+        flow_states.update_registration_flow(caller, flow_state)
+    })
+    // The update_registration_flow triggers a clean-up as well, which could lead to a situation where
+    // the flow being processed here is cleaned up. In that case, abort with an error.
+    .map_err(|_| CheckCaptchaError::NoRegistrationFlow)?;
+
+    Ok(IdRegNextStepResult {
+        next_step: RegistrationFlowNextStep::Finish,
+    })
+}
+
+pub fn identity_registration_finish(
+    arg: IdRegFinishArg,
+) -> Result<IdRegFinishResult, IdRegFinishError> {
+    let Some(current_state) = state::with_flow_states(|s| s.registration_flow_state(&caller()))
+    else {
+        return Err(IdRegFinishError::NoRegistrationFlow);
+    };
+
+    let RegistrationFlowState::FinishRegistration { .. } = current_state else {
+        return Err(IdRegFinishError::UnexpectedCall {
+            next_step: RegistrationFlowNextStep::from(current_state),
+        });
+    };
+
+    let identity_number = create_identity(&arg)?;
+
+    let caller = caller();
+    // flow completed --> remove flow state
+    state::with_flow_states_mut(|flow_states| flow_states.remove_registration_flow(&caller));
+
+    // add temp key so the user can keep using the identity used for the registration flow
+    state::with_temp_keys_mut(|temp_keys| {
+        temp_keys.add_temp_key(
+            TempKeyId::IdentityAuthnMethod {
+                identity_number,
+                authn_method_pubkey: arg.authn_method.public_key(),
+            },
+            caller,
+        )
+    });
+
+    Ok(IdRegFinishResult { identity_number })
+}
+
+fn create_identity(arg: &IdRegFinishArg) -> Result<IdentityNumber, IdRegFinishError> {
+    let Some(mut identity) = state::storage_borrow_mut(|s| s.allocate_anchor()) else {
```

Comment:
Is this a thing we will need to change to allow creating the identity number at the beginning of the flow instead of the end?

Reply:
Yes, there will definitely be changes here. But there need to be many more. Maybe its best we have a separate discussion about it.